modules/dnn/src/layers/nary_eltwise_layers.cpp

```diff
@@ -0,0 +1,427 @@
+            for (k = 0; k < narrays; k++) {
+                step[k][i] = 0;
+                shape[k][i] = 1;
+            }
+        }
+        return true;
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                         const int requiredOutputs,
+                         std::vector<MatShape> &outputs,
+                         std::vector<MatShape> &internals) const CV_OVERRIDE
+    {
+        MatShape outShape = findCommonShape(inputs);
+        outputs.assign(1, outShape);
+        return false;
+    }
+
+    template <typename T, typename Functor>
+    void do_broadcast_op(
+            int ndims, const int* shape,
+            const char* data1, const size_t* step1,
+            const char* data2, const size_t* step2,
+            char* data, const size_t* step,
+            const Functor& op)
+    {
+        assert(ndims >= 2);
+        size_t dp1 = step1[ndims-1]/sizeof(T);
+        size_t dp2 = step2[ndims-1]/sizeof(T);
+        size_t dp = step[ndims-1]/sizeof(T);
+        int k, n1 = shape[ndims-1], n2 = shape[ndims-2];
+        size_t plane_idx, nplanes = 1;
+        for (k = 0; k < ndims-2; k++) nplanes *= shape[k];
+
+        for (plane_idx = 0; plane_idx < nplanes; plane_idx++) {
+            const char* ptr1_ = data1;
+            const char* ptr2_ = data2;
+            char* ptr_ = data;
+            size_t idx = plane_idx;
+            for (k = ndims-3; k >= 0; k--) {
+                size_t next_idx = idx/shape[k];
+                int i_k = (int)(idx - next_idx*shape[k]);
+                ptr1_ += i_k*step1[k];
+                ptr2_ += i_k*step2[k];
+                ptr_ += i_k*step[k];
+                idx = next_idx;
+            }
+            for (int i2 = 0; i2 < n2; i2++, ptr1_ += step1[ndims-2],
+                                            ptr2_ += step2[ndims-2],
+                                            ptr_ += step[ndims-2])
+            {
+                const T* ptr1 = (const T*)ptr1_;
+                const T* ptr2 = (const T*)ptr2_;
+                T* ptr = (T*)ptr_;
+                if (dp1 == 1 && dp2 == 1 && dp == 1) {
+                    for(int i1 = 0; i1 < n1; i1++)
+                        ptr[i1] = op(ptr1[i1], ptr2[i1]);
+                } else if (dp1 == 1 && dp2 == 0 && dp == 1){
+                    T x2 = *ptr2;
+                    for(int i1 = 0; i1 < n1; i1++)
+                        ptr[i1] = op(ptr1[i1], x2);
+                } else if (dp1 == 0 && dp2 == 1 && dp == 1){
+                    T x1 = *ptr1;
+                    for(int i1 = 0; i1 < n1; i1++)
+                        ptr[i1] = op(x1, ptr2[i1]);
+                } else {
+                    for(int i1 = 0; i1 < n1; i1++, ptr1 += dp1, ptr2 += dp2, ptr += dp)
+                        *ptr = op(*ptr1, *ptr2);
+                }
+            }
+        }
+    }
+
+    template <typename T, typename Functor>
+    void forward_impl(const Functor& f, const Mat& a, const Mat& b, Mat& out)
+    {
```

Comment:
the first part of forward_impl is populated `number_of_elemwise_ops x number_of_types` times, however, it's enough to call it once before typeDispatch once and call `do_broadcast_op()` directly from `opDispatch()`. The only parameterization to call prepare_for_broadcast_op() is sizeof(T), however, it can be retrieved as `{a.elemSize(), b.elemSize(), out.elemSize()}`.

Reply:
I'll move this function to finalize(), since we don't need to reallocate and refold shapes every time.